### DISCUSSION TITLE
v3.4-beta1 - 2022-11-16

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#10941] [Announcements]
!!! warning "PostgreSQL 11 Required"
    NetBox v3.4 requires PostgreSQL 11 or later.

## Breaking Changes

* Device and virtual machine names are no longer case-sensitive. Attempting to create e.g. "device1" and "DEVICE1" will raise a validation error.
* The `asn` field has been removed from the provider model. Please replicate any provider ASN assignments to the ASN model introduced in NetBox v3.1 prior to upgrading.
* The `noc_contact`, `admin_contact`, and `portal_url` fields have been removed from the provider model. Please replicate any data remaining in these fields to the contact model introduced in NetBox v3.1 prior to upgrading.
* The `content_type` field on the CustomLink and ExportTemplate models have been renamed to `content_types` and now supports the assignment of multiple content types.
* The `cf` property on an object with custom fields now returns deserialized values. For example, a custom field referencing an object will return the object instance rather than its numeric ID. To access the raw serialized values, use `custom_field_data` instead.
* The `NetBoxModelCSVForm` class has been renamed to `NetBoxModelImportForm`. Backward compatability with the previous name has been retained for this release, but will be dropped in NetBox v3.5.

### New Global Search ([#10560](https://github.com/netbox-community/netbox/issues/10560))

NetBox's global search functionality has been completely overhauled and replaced by a new cache-based lookup. This new implementation provides a much speedier, more intelligent search capability. Matches are returned in order of precedence regardless of object type, and matched field values are highlighted in the results. Additionally, custom field values are now included in global search results (when enabled). Plugins can also register their own models with the new global search engine.

### Virtual Device Contexts ([#7854](https://github.com/netbox-community/netbox/issues/7854))

A new model representing virtual device contexts (VDCs) has been added. VDCs are logical partitions of resources within a device that can be managed independently. A VDC is created within a device and may have device interfaces assigned to it. An interface can be allocated to any number of VDCs on its device.

### Saved Filters ([#9623](https://github.com/netbox-community/netbox/issues/9623))

Object lists can be filtered by a variety of different fields and characteristics. Applied filters can now be saved for reuse as a convenience. Saved filters can be kept private, or shared among NetBox users.

### JSON/YAML Bulk Imports ([#4347](https://github.com/netbox-community/netbox/issues/4347))

NetBox's bulk import feature, which was previously limited to CSV-formatted data for most objects, has been extended to support the import of objects from JSON and/or YAML data as well.

### CSV-Based Bulk Updates ([#7961](https://github.com/netbox-community/netbox/issues/7961))

NetBox's CSV-based bulk import functionality has been extended to support also modifying existing objects. When an `id` column is present in the import form, it will be used to infer the object to be modified, rather than a new object being created. All fields (columns) are optional when modifying existing objects.

### Scheduled Reports & Scripts ([#8366](https://github.com/netbox-community/netbox/issues/8366))

Reports and custom scripts can now be scheduled for execution at a desired time.

### API for Staged Changes ([#10851](https://github.com/netbox-community/netbox/issues/10851))

This release introduces a new programmatic API that enables plugins and custom scripts to prepare changes in NetBox without actually committing them to the active database. To stage changes, create and activate a branch using the `checkout()` context manager. Any changes made within this context will be captured, recorded, and rolled back for future use. Once ready, a branch can be applied to the active database by calling `merge()`. 

!!! danger "Experimental Feature"
    This feature is still under active development and considered experimental in nature. Its use in production is strongly discouraged at this time.

## Enhancements

* [#6003](https://github.com/netbox-community/netbox/issues/6003) - Enable the inclusion of custom field values in global search
* [#7376](https://github.com/netbox-community/netbox/issues/7376) - Enable the assignment of tags during CSV import
* [#8245](https://github.com/netbox-community/netbox/issues/8245) - Enable GraphQL filtering of related objects
* [#8274](https://github.com/netbox-community/netbox/issues/8274) - Enable associating a custom link with multiple object types
* [#8485](https://github.com/netbox-community/netbox/issues/8485) - Enable journaling for all organizational models
* [#8853](https://github.com/netbox-community/netbox/issues/8853) - Introduce the `ALLOW_TOKEN_RETRIEVAL` config parameter to restrict the display of API tokens
* [#9249](https://github.com/netbox-community/netbox/issues/9249) - Device and virtual machine names are no longer case-sensitive
* [#9478](https://github.com/netbox-community/netbox/issues/9478) - Add `link_peers` field to GraphQL types for cabled objects
* [#9654](https://github.com/netbox-community/netbox/issues/9654) - Add `weight` field to racks, device types, and module types
* [#9817](https://github.com/netbox-community/netbox/issues/9817) - Add `assigned_object` field to GraphQL type for IP addresses and L2VPN terminations
* [#9832](https://github.com/netbox-community/netbox/issues/9832) - Add `mounting_depth` field to rack model
* [#9892](https://github.com/netbox-community/netbox/issues/9892) - Add optional `name` field for FHRP groups
* [#10052](https://github.com/netbox-community/netbox/issues/10052) - The `cf` attribute now returns deserialized custom field data
* [#10348](https://github.com/netbox-community/netbox/issues/10348) - Add decimal custom field type
* [#10545](https://github.com/netbox-community/netbox/issues/10545) - Standardize the use of `description` and `comments` fields on all primary models
* [#10556](https://github.com/netbox-community/netbox/issues/10556) - Include a `display` field in all GraphQL object types
* [#10595](https://github.com/netbox-community/netbox/issues/10595) - Add GraphQL relationships for additional generic foreign key fields
* [#10698](https://github.com/netbox-community/netbox/issues/10698) - Omit app label from content type in table columns
* [#10710](https://github.com/netbox-community/netbox/issues/10710) - Add `status` field to WirelessLAN
* [#10761](https://github.com/netbox-community/netbox/issues/10761) - Enable associating an export template with multiple object types
* [#10781](https://github.com/netbox-community/netbox/issues/10781) - Add support for Python v3.11

## Plugins API

* [#4751](https://github.com/netbox-community/netbox/issues/4751) - Add `plugin_list_buttons` template tag to embed buttons on object lists
* [#8927](https://github.com/netbox-community/netbox/issues/8927) - Enable inclusion of plugin models in global search via `SearchIndex`
* [#9071](https://github.com/netbox-community/netbox/issues/9071) - Enable plugins to register top-level navigation menus
* [#9072](https://github.com/netbox-community/netbox/issues/9072) - Enable registration of tabbed plugin views for core NetBox models
* [#9880](https://github.com/netbox-community/netbox/issues/9880) - Enable plugins to install and register other Django apps
* [#9887](https://github.com/netbox-community/netbox/issues/9887) - Inspect `docs_url` property to determine link to model documentation
* [#10314](https://github.com/netbox-community/netbox/issues/10314) - Move `clone()` method from NetBoxModel to CloningMixin
* [#10543](https://github.com/netbox-community/netbox/issues/10543) - Introduce `get_plugin_config()` utility function
* [#10739](https://github.com/netbox-community/netbox/issues/10739) - Introduce `get_queryset()` method on generic views

## Other Changes

* [#9045](https://github.com/netbox-community/netbox/issues/9045) - Remove legacy ASN field from provider model
* [#9046](https://github.com/netbox-community/netbox/issues/9046) - Remove legacy contact fields from provider model
* [#10358](https://github.com/netbox-community/netbox/issues/10358) - Raise minimum required PostgreSQL version from 10 to 11
* [#10694](https://github.com/netbox-community/netbox/issues/10694) - Emit the `post_save` signal when creating device components in bulk
* [#10697](https://github.com/netbox-community/netbox/issues/10697) - Move application registry into core app
* [#10699](https://github.com/netbox-community/netbox/issues/10699) - Remove custom `import_object()` function
* [#10816](https://github.com/netbox-community/netbox/issues/10816) - Pass the current request when instantiating a FilterSet within UI views
* [#10820](https://github.com/netbox-community/netbox/issues/10820) - Switch timezone library from pytz to zoneinfo
* [#10821](https://github.com/netbox-community/netbox/issues/10821) - Enable data localization

<hr /><em>This discussion was created from the release <a href='https://github.com/netbox-community/netbox/releases/tag/v3.4-beta1'>v3.4-beta1 - 2022-11-16</a>.</em>

**Comment by jnovak-netsystemcz** [comment]
Hi. I read announcement here: https://netbox.dev/categories/#announcements
There is mentioned public demo https://beta-demo.netbox.dev/ with admin/admin. The server is reachable, but described username/password doesn't work.
Can you clarify which username/password I should use for testing the beta?

**Reply by candlerb** [reply]
admin/admin works for me (with macOS and Chrome).  Maybe it's been fixed?

If it still doesn't work for you, then please show the exact error displayed, and also check your browser console for errors.  It could be your browser not accepting cookies, or some content filtering or anti-virus tool you have installed.

**Reply by jnovak-netsystemcz** [reply]
Thank you for reply. I retested it a few minutes ago with no success.
I see 'Please enter a correct username and password. Note that both fields may be case-sensitive.' error and I'm using 'admin' as user and 'admin' as password (which works on stable demo). So I quite confident I'm using correct credentials.
I tested it with FF and Chrome from my PC and with Chrome from my mobile device connected with completely different provided and nothing helps.
There is no error on browser console. I checked what I'm sending in POST and I see what I expect:

csrfmiddlewaretoken	"ZrKvDNFUCyCi8NzIcr5J0Dfr4mexJHy3lgecOhL4GcDTkWHt3GCNOMhhtit42Drd"
next	"/"
username	"admin"
password	"admin"

**Comment by hagbarddenstore** [comment]
It doesn’t work for me either. I’m guessing someone or something is changing the password. Perhaps Jeremy forgot to add the “disable admin password change”-plugin Sent from my iPhoneOn 29 Nov 2022, at 11:29, jnovak-netsystemcz ***@***.***> wrote:﻿
Thank you for reply. I retested it a few minutes ago with no success.
I see 'Please enter a correct username and password. Note that both fields may be case-sensitive.' error and I'm using 'admin' as user and 'admin' as password (which works on stable demo). So I quite confident I'm using correct credentials.
I tested it with FF and Chrome from my PC and with Chrome from my mobile device connected with completely different provided and nothing helps.
There is no error on browser console. I checked what I'm sending in POST and I see what I expect:
csrfmiddlewaretoken	"ZrKvDNFUCyCi8NzIcr5J0Dfr4mexJHy3lgecOhL4GcDTkWHt3GCNOMhhtit42Drd"
next	"/"
username	"admin"
password	"admin"

**Reply by candlerb** [reply]
It is failing for me now.  Perhaps it's something to do with the daily process that resets it to defaults?

**Reply by jeremystretch** [reply]
This has been resolved. Changes made to address bugs #9223 and #10969 inadvertently broke the population of the demo data set. Login was impossible because no users had been created. (These changes were actually made under the v3.3 branch and would have affected the v3.3 demo as well upon the release of v3.3.9 had they not been resolved.)

**Reply by jnovak-netsystemcz** [reply]
It works now. Thank you.

**Comment by vitalkinum** [comment]
Good day, when you open some devices, you get this error



Server Error
There was a problem with your request. Please contact an administrator.

The complete exception is provided below:

<class 'AttributeError'>

'NoneType' object has no attribute 'napalm_driver'

Python version: 3.8.10
NetBox version: 3.4-beta1

**Reply by candlerb** [reply]
What *exactly* did you do? And what was in the URL bar in your browser at the time of the exception?

**Reply by kkthxbye-code** [reply]
It's already fixed, will be included in beta2 or final release. You can work around it for testing it by setting a platform for the device or removing the primary IP.

**Reply by vitalkinum** [reply]
> It's already fixed, will be included in beta2 or final release. You can work around it for testing it by setting a platform for the device or removing the primary IP.


thanks, it helped

**Comment by ravenrs** [comment]
Hi, I'm trying to upgrade from 3.1.5 to 4.0.9 through  3.7.8, it should be possible according to the manual. But I've got a problem:
```
django.db.utils.DataError: Unable to proceed with deleting any field from Provider model: Found 235 providers with legacy ASN data. Please ensure all legacy provider ASN data has been migrated to AS objects before proceeding.
 Or, set the NETBOX_DELETE_LEGACY_DATA environment variable to bypass this safeguard and delete all legacy provider ASN data.
```
I saw a comment: 
`The asn field has been removed from the provider model. Please replicate any provider ASN assignments to the ASN model introduced in NetBox v3.1 prior to upgrading. `
But it's impossible because the provider class in the provider model in 3.1.5 has no ASNs fields for example, it has a site class.class 
```

Provider(PrimaryModel):
    """
    Each Circuit belongs to a Provider. This is usually a telecommunications company or similar organization. This model
    stores information pertinent to the user's relationship with the Provider.
    """
    name = models.CharField(
        max_length=100,
        unique=True
    )
    slug = models.SlugField(
        max_length=100,
        unique=True
    )
    asn = ASNField(
        blank=True,
        null=True,
        verbose_name='ASN',
        help_text='32-bit autonomous system number'
    )
    account = models.CharField(
        max_length=30,
        blank=True,
        verbose_name='Account number'
    )
    portal_url = models.URLField(
        blank=True,
        verbose_name='Portal URL'
    )
    noc_contact = models.TextField(
        blank=True,
        verbose_name='NOC contact'
    )
    admin_contact = models.TextField(
        blank=True,
        verbose_name='Admin contact'
    )
    comments = models.TextField(
        blank=True
    )

    # Generic relations
    contacts = GenericRelation(
        to='tenancy.ContactAssignment'
    )

    clone_fields = [
        'asn', 'account', 'portal_url', 'noc_contact', 'admin_contact',
    ]

    class Meta:
        ordering = ['name']

    def __str__(self):
        return self.name

    def get_absolute_url(self):
        return reverse('circuits:provider', args=[self.pk])
```

So I need to store this info and restore it only when I upgrade to 4.0.9. It's not obvious from release not, or you should change the migration script and add a migration function as you did in 0042_provideraccount.py of 3.7.8 version.
